src/vs/workbench/contrib/search/browser/notebookSearch/notebookSearchService.ts

```diff
@@ -0,0 +1,212 @@
+					return <ISearchComplete>{
+						messages: [],
+						limitHit: resolved.reduce((prev, cur) => prev || cur.limitHit, false),
+						results,
+					};
+				}),
+				allScannedFiles: promise.then(resolvedPromise => {
+					const openNotebookResults = resolvedPromise[0];
+					const closedNotebookResults = resolvedPromise[1];
+					const results = arrays.coalesce([...openNotebookResults.results.keys(), ...closedNotebookResults?.results.keys() ?? []]);
+					return new ResourceSet(results, uri => this.uriIdentityService.extUri.getComparisonKey(uri));
+				})
+			};
+		};
+		const promiseResults = getAllResults();
+		return {
+			openFilesToScan: new ResourceSet(localNotebookFiles),
+			completeData: promiseResults.completeData,
+			allScannedFiles: promiseResults.allScannedFiles
+		};
+	}
+
+	private async getClosedNotebookResults(textQuery: ITextQuery, scannedFiles: ResourceSet, token: CancellationToken): Promise<IClosedNotebookSearchResults> {
+
+		const sortedContributedTypes = [...this.notebookService.getContributedNotebookTypes()].sort((a, b) => priorityToRank(b.priority) - priorityToRank(a.priority));
+
+		const getResultsFromProviderInfo = async (providerInfo: NotebookProviderInfo) => {
+			const serializer = (await this.notebookService.withNotebookDataProvider(providerInfo.id)).serializer;
+			return await serializer.searchInNotebooks(textQuery, token);
+		};
+
+		const start = Date.now();
+		const searchComplete = (await Promise.all(sortedContributedTypes.map(async e => await getResultsFromProviderInfo(e))));
+		const results = searchComplete.flatMap(e => e.results);
+		let limitHit = searchComplete.some(e => e.limitHit);
+
+		// results are already sorted with high priority first, filter out duplicates.
+		const uniqueResults = new ResourceMap<INotebookFileMatchNoModel | null>(uri => this.uriIdentityService.extUri.getComparisonKey(uri));
+
+		let numResults = 0;
+		for (const result of results) {
+			if (textQuery.maxResults && numResults >= textQuery.maxResults) {
+				limitHit = true;
+				break;
+			}
+
+			if (!scannedFiles.has(result.resource) && !uniqueResults.has(result.resource)) {
+				uniqueResults.set(result.resource, result.cellResults.length > 0 ? result : null);
+				numResults++;
+			}
+		}
+
+		const end = Date.now();
+		this.logService.trace(`query: ${textQuery.contentPattern.pattern}`);
+		this.logService.trace(`closed notebook search time | ${end - start}ms`);
+
+		return {
+			results: uniqueResults,
+			limitHit
+		};
+	}
+
+	private async getLocalNotebookResults(query: ITextQuery, token: CancellationToken, widgets: Array<NotebookEditorWidget>, searchID: string): Promise<IOpenNotebookSearchResults> {
+		const localResults = new ResourceMap<INotebookFileMatchWithModel | null>(uri => this.uriIdentityService.extUri.getComparisonKey(uri));
+		let limitHit = false;
+
+		for (const widget of widgets) {
+			if (!widget.viewModel) {
```

Comment:
techinically it's the same as `widget.hasModel()`, we might want to avoid using `viewModel` if possible.